master/static/migrations/20240112140933_add-storage-config.tx.up.sql

```diff
@@ -0,0 +1,190 @@
+CREATE TABLE storage_backend_shared_fs (
+  id integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
+  host_path        TEXT NOT NULL,
+  container_path   TEXT,
+  checkpoint_path  TEXT,
+  tensorboard_path TEXT,
+  storage_path     TEXT,
+  propagation      TEXT NOT NULL
+  CONSTRAINT shared_fs_reserved_value CHECK (
+    container_path != 'DeterminedReservedNullUniqueValue' AND
+    checkpoint_path != 'DeterminedReservedNullUniqueValue' AND
+    tensorboard_path != 'DeterminedReservedNullUniqueValue' AND
+    storage_path != 'DeterminedReservedNullUniqueValue'
+  )
+);
+
+-- Hack for lack of pre-postgres 15 NULLS NOT DISTINCT.
+-- https://stackoverflow.com/questions/8289100/create-unique-constraint-with-null-columns
+CREATE UNIQUE INDEX ix_storage_backend_unique_fs ON storage_backend_shared_fs (
+  host_path,
+  COALESCE(container_path, 'DeterminedReservedNullUniqueValue'),
+  COALESCE(checkpoint_path, 'DeterminedReservedNullUniqueValue'),
+  COALESCE(tensorboard_path, 'DeterminedReservedNullUniqueValue'),
+  COALESCE(storage_path, 'DeterminedReservedNullUniqueValue'),
+  propagation
+);
+
+
+CREATE TABLE storage_backend_s3 (
+  id integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
+  bucket       TEXT NOT NULL,
+  access_key   TEXT,
+  secret_key   TEXT,
+  endpoint_url TEXT,
+  prefix       TEXT,
+  CONSTRAINT s3_check_prefix CHECK (
```

Comment:
thoughts are mirroring the expconf logic into the database?

I am worried somewhat that this logic will cause a bug or it will get out of sync so I can drop these checks. I am also worried that without them eventually someone will write a migration or insert without validating it?

Reply:
These checks written in Go should also happen in the ```schemas.IsComplete``` before we insert it.

Doing them in the database makes sure they must have happened in Go before insert and someone didn't skip it

It also protects against bad migrations

Reply:
okay; then this doesn't sound like a concern. if we need to take it out we'd still have go level validation.